### PR TITLE
cached iterator

### PR DESCRIPTION
**Pull request by ibeltagy**
A solution for this issue: https://github.com/allenai/allennlp/issues/1128

I implemented it as a subclass of `BucketIterator` just to avoid messing up other stuff. However, it is better to be implemented as an additional option in `BasicIterator` where it is only enabled when `dataset_reader` is not lazy and `max_instances_in_memory is None`.

### CODE REVIEW

File: allennlp/data/iterators/cached_iterator.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    An extension of the `BucketIterator` that caches its batches between epochs. Batches are built
+    only once during the first epoch, then the same batches are used again in consecutive epochs.
+    Using this iterator saves run time but costs more memory to store the cached batches. It should
+    be used only if the dataset can fit in memory. This iterator shouldn't be used with the lazy
+    mode.
+
+    Parameters:
+    ----------
+        See :class:`BucketIterator`.
+    """
```
Comment (joelgrus) [inline allennlp/data/iterators/cached_iterator.py]: I would implement a constructor that throws a `ConfigurationError` if you try to instantiate it with lazy=True

File: allennlp/data/iterators/cached_iterator.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    @overrides
+    def _yield_one_epoch(self, instances: Iterable[Instance], shuffle: bool, cuda_device: int, for_training: bool):
+        if not hasattr(self, 'cached_batches'):
+            # instances id -> list[batches]
+            self.cached_batches: Dict[int, List] = {}  # pylint: disable=attribute-defined-outside-init
```
Comment (joelgrus) [inline allennlp/data/iterators/cached_iterator.py]: and then I would just define this in the constructor

Comment (ibeltagy) [reply allennlp/data/iterators/cached_iterator.py]: do you know how to define a constructor for a subclass that takes the same arguments of the parent class without explicitly listing them ? I don't want to copy-paste them from `BucketIterator.__init__`

Comment (DeNeutoy) [reply allennlp/data/iterators/cached_iterator.py]:
```
class SubClass( Super ):
   def __init__(self, *args, **kwargs):
       super(self).__init__(*args, **kwargs)

```

File: allennlp/data/iterators/cached_iterator.py (hunk 1 above)
Comment (joelgrus) [reply allennlp/data/iterators/cached_iterator.py]: although actually, laziness is a function of the dataset, not of the iterator, so this doesn't make sense

**Comment by ibeltagy** [comment]
@joelgrus, added the check for laziness, and yes, I agree, this fits in a class higher in the hierarchy. `BasicIterator` would be a better fit than `DataIterator` because that's where `max_instances_in_memory` was introduced. 
Will let you guys decide if you want to keep it in its own class or somewhere else.

**Comment by ibeltagy** [comment]
This PR still needs more work to make sure batches are cached in main memory not gpu memory

**Comment by ibeltagy** [comment]
This is working fine now. An example is here: http://beaker.allenai.org/ex/ex_gnghsslk14vz/tasks. Duration of the first epoch is 42min then drops to 28min in the subsequent ones.

**Comment by schmmd** [comment]
@joelgrus what is the status of this PR now?  Is this something we would like to merge?  If not, why are we no longer interested in it?

**Comment by joelgrus** [comment]
this functionality was added to the base `DataIterator` class, so this pr is no longer necessary

**Pull request closed without merging**